btcec/nfield_test.go

```diff
@@ -0,0 +1,925 @@
+		in2      string // second hex encoded value to multiply with
+		expected string // expected hex encoded value
+	}{
+		{"0", "0", "0"},
+		{"1", "0", "0"},
+		{"0", "1", "0"},
+		{"1", "1", "1"},
+		// secp256k1 n-1 * 2
+		{
+			"fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364140",
+			"2",
+			"fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd036413f",
+		},
+		// secp256k1 n * 3
+		{
+			"fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364141",
+			"3",
+			"0",
+		},
+		// secp256k1 n-1 * 8
+		{
+			"fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364140",
+			"8",
+			"fffffffffffffffffffffffffffffffebaaedce6af48a03bbfd25e8cd0364139",
+		},
+		// Random samples.
+		{
+			"cfb81753d5ef499a98ecc04c62cb7768c2e4f1740032946db1c12e405248137e",
+			"58f355ad27b4d75fb7db0442452e732c436c1f7c5a7c4e214fa9cc031426a7d3",
+			"7a3e56db90348454f42571063ccc9a13956293a14ba03c07418b11ab36c0086d",
+		},
+		{
+			"26e9d61d1cdf3920e9928e85fa3df3e7556ef9ab1d14ec56d8b4fc8ed37235bf",
+			"2dfc4bbe537afee979c644f8c97b31e58be5296d6dbc460091eae630c98511cf",
+			"f8592ec43b9f42b914be9ec641a89a58ad0d62c00e8d86640559b8d5ae91d33f",
+		},
+		{
+			"5db64ed5afb71646c8b231585d5b2bf7e628590154e0854c4c29920b999ff351",
+			"279cfae5eea5d09ade8e6a7409182f9de40981bc31c84c3d3dfe1d933f152e9a",
+			"89f90fd64b0d7579444dfbb2745dfdda883893f848f5743faca560901d8b89dd",
+		},
+		{
+			"b66dfc1f96820b07d2bdbd559c19319a3a73c97ceb7b3d662f4fe75ecb6819e6",
+			"bf774aba43e3e49eb63a6e18037d1118152568f1a3ac4ec8b89aeb6ff8008ae1",
+			"5a2590e74896908bf295a7ee473e06048c961c4fbe39d812ab0298ecd3241415",
+		},
+	}
+
+	t.Logf("Running %d tests", len(tests))
+	for i, test := range tests {
+		f := btcec.NewNfieldVal().SetHex(test.in1).Normalize()
+		f2 := btcec.NewNfieldVal().SetHex(test.in2).Normalize()
+		expected := btcec.NewNfieldVal().SetHex(test.expected).Normalize()
+		result := f.Mul(f2).Normalize()
+		if !result.Equals(expected) {
+			t.Errorf("nfieldVal.Mul #%d wrong result\n"+
+				"got: %v\nwant: %v", i, result, expected)
+			continue
+		}
+	}
+}
+
+// TestNfieldMulRand tests the multiplication correctness by doing the
+// same calculation with Big.Int
+func TestNfieldMulRand(t *testing.T) {
+	data := make([]byte, 32)
+	_, err := rand.Read(data)
+	if err != nil {
+		t.Fatalf("failed to read random data")
+	}
+	N := btcec.S256().N
+	a := new(big.Int).SetBytes(data)
+	aN := btcec.NewNfieldVal().SetByteSlice(a.Bytes())
+	for i := 0; i < 100; i++ {
+		data = make([]byte, 32)
+		_, err := rand.Read(data)
+		if err != nil {
+			t.Errorf("failed to read random data")
+			continue
+		}
+		b := new(big.Int).SetBytes(data)
+		c := big.NewInt(0)
+		c.Mul(a, b)
+		c.Mod(c, N)
+		bN := btcec.NewNfieldVal().SetByteSlice(b.Bytes())
+		cN := btcec.NewNfieldVal()
+		cN.Mul2(aN, bN)
+		cstr := fmt.Sprintf("%064x", c)
+		if cN.String() != cstr {
+			t.Fatalf("expected strings to be the same, was not: %v %v", cN.String(), cstr)
```

Comment:
I would highly suggest this test pick a random seed to use for the random number generation and log it in the event of failure.  Alternatively, this test should log the operands used in the calculation.

With neither of those things, if there is a failure, there would be no way to reproduce it in order to debug it.